typescript: Add typescript support.

vsvipul opened the pull request:
Adds typescript support using tsc for both main and renderer process.

Need to figure out the lint-js part using xo.

**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

priyank-p commented on tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
We want to use `esnext` here because we don't want the javascript files to be transpliled for older version, electron always supports newer javascript syntax.

vsvipul replied:
ok sure!

priyank-p commented on tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "strict": true,                           /* Enable all strict type-checking options. */
+    "esModuleInterop": true,                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "noImplicitAny": false
```
This should be `true` to make sure we take full advantage of typescript and let it complain about where we should do type annotations.

vsvipul replied:
ok gotcha !

priyank-p commented on typings.d.ts:
```diff
@@ -0,0 +1 @@
+declare module '*';
```
Why is this required? The reason I am asking is to know what problem you faced (I know that we need to do something like this to make this work).

vsvipul replied:
This is to avoid compiler warning for having declaration files for each and every util we use.

priyank-p commented on app/main/menu.ts:
```diff
@@ -211,7 +211,7 @@ class AppMenu {
 
 	getWindowSubmenu(tabs, activeTabIndex) {
-		const initialSubmenu = [{
+		const initialSubmenu: any[] = [{
```
We should declare a interface for this. Using a interface would  will make sure any function consuming this will get exactly what it expected.

(The `noImplicitAny` tsconfig.json option will make sure we can't do this.)

vsvipul replied:
I changed noImplicitAny to ```true``` but still could do this.

priyank-p reviewed (commented):
@vsvipul can you split the PR into coherent commits, like each individual commits for: adding dependencies, adding tsconfig.json, migration a file (one commit per file), and for setting up the xo-typescript.

I am looking into why xo-typescript doesn't work!

priyank-p commented:
@vsvipul I think the xo-typescript thing is not working because you are missing `eslint-config-xo`; the documentation tells us to install that and I think that's what is responsible for adding xo-typescript config to xo. Did not have a chance to test it out yet though :/

vsvipul commented:
@priyank-p i installed everything mentioned in the documentation.  xo-typescript is working currently but no t with the old js rules, it is using some new rules and some new errors are popping up like these - 
![Screenshot from 2019-06-16 10-06-34](https://user-images.githubusercontent.com/21038781/59559121-7ce03d80-901e-11e9-8fb6-81296be55e12.png)

vsvipul committed:
dependencies: Add dependencies for typescript.

Add dependencies and sets up typescript
and xo for typescript.

vsvipul committed:
Add tsconfig.json and global typings.

Adds tsconfig.json and a typings.d.ts file
which is required to avoid compiler warnings for
having declaration for each and every util
we use.

vsvipul committed:
typescript: Migrate menu.js to typescript.

vsvipul committed:
typescript: Migrate startup.js to typescript.

vsvipul commented:
@priyank-p I have split the commit to multiple commits. Working on migrating other files now.

vsvipul committed:
typescript: Migrate system-util.js to typescript.

We use to check for a case where os.platform() == win64
which is not possible so removed the condition.

vsvipul committed:
typescript: Migrate sentry-util.js to typescript.

vsvipul committed:
typescript: Migrate request-util to typescript.

We use newer javascript method for string that linting enforces like
startsWith, includes.

vsvipul committed:
typescript: Migrate reconnect-util to typescript.

vsvipul committed:
typescript: Migrate config-util to typescript.

vsvipul committed:
typescript: Migrate logger-util to typescript.

akashnimare commented on app/main/menu.ts:
```diff
@@ -1,15 +1,15 @@
-const DNDUtil = require(__dirname + '/../renderer/js/utils/dnd-util.js');
-const Logger = require(__dirname + '/../renderer/js/utils/logger-util.js');
+import { ConfigUtil } from '../renderer/js/utils/config-util';
+import DNDUtil from '../renderer/js/utils/dnd-util';
```
{ DNDUtil } ?

vsvipul replied:
That hasn't been migrated to typescript for now. So, for the sake of testing, left it as it was before. Will change when i migrate DNDUtil to typescript.

akashnimare replied:
Okay, sounds good.

akashnimare commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -32,7 +32,7 @@ class RequestUtil {
 			// To handle case where certificate has been moved from the location in certificates.json
 			try {
-				certificateLocation = fs.readFileSync(certificate);
+				certificateLocation = fs.readFileSync(certificate).toString();
```
Is this needed, `toString()`?

priyank-p replied:
Yes, you need to do that otherwise this would be a buffer. Another way which is think makes more sense is to pass `utf8` as the encoding parameter.

akashnimare commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -1,12 +1,17 @@
+}
 
-class SystemUtil {
+export class systemUtil {
```
@priyank-p @vsvipul I don't think export is needed here, can be removed.

priyank-p replied:
Fixed this.

akashnimare commented on package.json:
```diff
@@ -142,19 +145,26 @@
+    "typescript": "^3.5.2",
+    "xo": "^0.24.0"
   },
   "xo": {
```
@priyank-p, @vsvipul do we have an option to put the `xo` config in a separate file since it's getting longer.

priyank-p replied:
No, `xo` doesn't provide a way to use a config file. Though we can use a `.eslintrc` I think.

akashnimare commented on package.json:
```diff
@@ -180,12 +190,18 @@
+  },
+  "dependencies": {
+    "@types/node": "^12.0.8",
+    "@types/requestidlecallback": "^0.1.1"
```
@vsvipul not sure what's this?

priyank-p replied:
This was added because @vsvipul was getting a error when he migrated a ts that used this function. I will look into this.

priyank-p committed:
dependencies: Add @types/node.

priyank-p committed:
typescript: Run typescript compiler when running npm start.

We use concurrently to run both electron app and tsc -w toghether
in parallel.

priyank-p commented:
@vsvipul I fixed the linting errors we were getting out of nowhere by upgrading `xo` to `v0.22.0`. It is pretty weird that by just upgrading `xo` we would start to get errors we should look into this, I think the eslint upgrade is what's causing this. Note that there are still linting error in the PR, but these are errors we'll want to fix and are coming from eslint-typescript plugin.

I merged the dependencies (into typescript branch) and tsconfig.json commits with some tweaks. I also added a commit which runs `tsc -w` when we run `npm start` still working on `gulp` part of this.

priyank-p commented:
(Also, I like to mention if you checkout the PR, you will need to do `git clean -d` otherwise `git` won't let you switch branches.)

priyank-p committed:
dependencies: Downgrade @types/node to v10.1.0.

This has to done because we use node v10.2.0 not the latest
version of node. There is v10.2.0 for @types/node so we use
something closer to it.

priyank-p committed:
.gitignore: Ignore typescript compiled files.

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -1,19 +1,19 @@
+let instance: any = null;
 
-class RequestUtil {
+export class requestUtil {
```
@vsvipul why did you rename the class? It should be `RequestUtil`

vsvipul replied:
We needed to export an instance of RequestUtil to make it work in other files, otherwise tsc was returning an error. So renamed the class and named the instance RequestUtil.

priyank-p replied:
We can do that as `export { requestUtil as RequestUtil };`

vsvipul replied:
oh, will change. Thanks.

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -23,7 +23,7 @@ class RequestUtil {
 	// ignoreCerts parameter helps in fetching server icon and
 	// other server details when user chooses to ignore certificate warnings
-	requestOptions(domain, ignoreCerts) {
+	requestOptions(domain: string, ignoreCerts: boolean) {
```
I added a new interface here for annotating the return value of this function before merging this commit.

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -49,7 +49,7 @@ class RequestUtil {
 	}
 
-	formatUrl(domain) {
+	formatUrl(domain: string) {
```
Annotated return type here.

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -49,7 +49,7 @@ class RequestUtil {
 
-	formatUrl(domain) {
+	formatUrl(domain: string) {
 		const hasPrefix = (domain.indexOf('http') === 0);
```
The linter was complaining about using `startsWith` here so made that change here

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -1,19 +1,19 @@
 });
 
-let instance = null;
+let instance: any = null;
```
change this to `null | SystemUtil`

priyank-p commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -1,12 +1,17 @@
+import * as os from 'os';
 
-const os = require('os');
+let instance: any = null;
```
changed this to `null | SystemUtil`.

priyank-p commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -1,12 +1,17 @@
+}
 
-class SystemUtil {
+export class systemUtil {
```
Removed the interface and directly annoated the class.

priyank-p commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
 	}
 
-	setUserAgent(webViewUserAgent) {
+	setUserAgent(webViewUserAgent: any) {
```
this should be `string`

priyank-p commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
 			return 'Linux';
-		}
-		if (os.platform() === 'win32' || os.platform() === 'win64') {
+		} else if (platform === 'win32' || platform === 'win64') {
```
Removed the `win64` check

vsvipul replied:
win64 doesn't exist! So, was remove therefore. See https://nodejs.org/api/os.html#os_os_platform .

priyank-p commented on app/renderer/js/utils/system-util.ts:
```diff
@@ -28,22 +33,23 @@ class SystemUtil {
 
 	getOS() {
-		if (os.platform() === 'darwin') {
+		const platform: string = os.platform();
```
I remove the `string` declaration there. The function return a type `NodeJS.Platform` which we want to use instead. Genral rule is you shouldn't annotate type that are implicitly annotated.

For example:
```typescript
// Do
const a: string;
a = 'aa';

// Don't
const a: string = 'aa';
```

vsvipul replied:
Got it! Will keep in mind from now on.

priyank-p commented on app/renderer/js/utils/request-util.ts:
```diff
@@ -1,19 +1,19 @@
-const fs = require('fs');
-const Logger = require('./logger-util');
+import * as fs from 'fs';
+import { Logger } from './logger-util';
```
This should be just `import Logger from './logger-util'`

priyank-p commented:
@vsvipul posted the thing I changed when merging the request-util and logger-util commits above.

priyank-p committed:
request-util: Fix linting error.

vsvipul committed:
typescript: Fixes to import statements

Fixes to import statements in files to make things
work.

vsvipul committed:
typescript: Migrate certificate-util to typescript.

vsvipul committed:
typescript: Migrate common-util to typescript.

vsvipul committed:
typescript: Migrate default-util to typescript.

vsvipul committed:
typescript: Migrate link-util to typescript.

vsvipul committed:
typescript: Migrate dnd-util to typescript.

priyank-p committed:
typescript: Fix wrong import syntax in request-util.ts.

Using the previous syntax was wrong because the file export
the class For example we would use previous syntax if the class
was export by `module.export = { ProxyUtil }`.

priyank-p committed:
typing.d.ts: Add comment for why we use declare module.

priyank-p committed:
xo: Disable lines-between-class-members rule.

Apparently this doesn't work well with typescript files.
For example this will throw this linting error:
class Test {
  a: boolean;
  b: string; // needs a new line before this one to fix this linting error
}

priyank-p commented:
Merged all of the commits with changes to the typescript branch. Thanks @vsvipul for working on this, we just finished migration of the whole `utils` directory which is awesome.

Next, let finish migrating the rest of the three files left in `app/main` directory.

abhigyank commented:
This is great work @vsvipul @priyank-p 🎉
